### ISSUE TITLE
How to get test predictions (and other non-scalar metrics)?

### ISSUE THREAD

**Issue by KeAWang** [PyTorchLightning/pytorch-lightning#1166]
#### What is your question?
If I have a trained model and I want to test it using `Trainer.test()`, how do I get the actual predictions of the model on the test set?

I tried to log the predictions and writing a Callback to get the logs at test end, but it seems like I can only log scalar Tensors in the dictionary returned by my model's `test_end()`.

**Comment by Borda** [comment]
Do need also to measure model performances? if not you can just get predictions, see
https://pytorch-lightning.readthedocs.io/en/latest/introduction_guide.html#predicting

**Comment by KeAWang** [comment]
It would be nice to not have to call the model separately since then I would have to write my own testing loop on the test set which kinda defeats the purpose of Pytorch lightning

**Comment by awaelchli** [comment]
@KeAWang 
If I understand correctly, you want to log ALL predictions on your test set. This should be possible:

- Implement `test_dataloader`, `test_step` and `test_end` 
- In `test_step` you can't return a log dict, that's normal. But you can call `self.logger.experiment.add_image` or whatever method you want to use to log your predictions (depends on the logger class). 

Is this what you need?

**Comment by KeAWang** [comment]
Right but I want to be able to get the predictions as a local variable that I can continue to process in my jupyter notebook. I'm not sure how to extract it from a tensorboard logger. Would I have to write my own logger?

**Comment by awaelchli** [comment]
That's not the same as logging, that was not clear in your original question.
You will have to collect your predictions in test_step in a variable like `self.predictions` which is a list or something. Then after you call `trainer.test()` you can access `model.predictions` in your notebook. 
What do you think?

**Comment by KeAWang** [comment]
I was wondering if there was something else that fit within the framework. But that would work. Thank you!

**Comment by manpreetsinghminhas** [comment]
If you want direct access, it can be done directly using the forward pass as follows. (The way it is done in PyTorch.)

```
model  = YourLightningModule.load_from_checkpoint(r"path/to/checkout.ckpt")
model.to(device)
data.to(device)
with torch.no_grad():
    out = model(data)
```
You can loop over a dataloader if you want.

**Comment by talhaanwarch** [comment]
> That's not the same as logging, that was not clear in your original question.
> You will have to collect your predictions in test_step in a variable like `self.predictions` which is a list or something. Then after you call `trainer.test()` you can access `model.predictions` in your notebook.
> What do you think?

i tried its empty

**Comment by yipliu** [comment]
Can anyone make it clear about the  **self.trainer.results**. This is useful in the following scenario.

1. Test dataset: all label is set to zero
2. The predictions (in test dataset ) have to align with the metadata
3.  All x, y, and metadata are processed in Dataset together

In this scenario, we just call **test_epoch_end**, it will return the prediction and aligned metadata
